### ISSUE TITLE
Make URL for websocket server configurable

### ISSUE THREAD

**Issue by bledo22** [kivle/msfs-map#1]
Sorry if it's the wrong channel and/or wrong time... Can I ask something about this add-on or is it closed for good? Thanks!

**Comment by kivle** [comment]
Hi,

Certainly, feel free to ask anything! :)

The project isn't really dead, per se. It should still work as it is, and more features have been added since I made that Youtube video to demo it.

I did however reach some limits to what I could do with this approach without doing some major rethink. These are the current problems with how it works:

1. In a lot of areas on the globe (eg. Africa, etc.) there simply isn't many geotagged articles, so you might be flying for an hour before something shows up.
2. In other areas (like the UK for instance), the density of geotagged articles is simply too large. If you fly over London, it will find 3-4 geotagged articles for every few kilometers you fly. So you end up with a huge backlog of articles that the TTS simply does not get the time to read. It will constantly keep reading new articles but the backlog keeps growing and growing.
3. In some areas in for instance central Europe there are a lot of "stub" articles for little villages. So you might fly over a small area and get 10 articles that are just "Village X is located in region Y of country Z. It has a population of C". That can get pretty boring, so it would be nice to find a way to filter these out.
4. When it searches for articles, the application currently only requests the summary. So it will only retrieve the first 100-200 words of an article. Generally the introduction to the full Wikipedia article. In most cases that's good enough, but in a lot of cases the more interesting stuff might be further down in the text of a Wikipedia article. But if it were to request the full article text, it would need some way of picking out only the most interesting information in an automated way.

So I tried various strategies of filtering out uninteresting articles. I tried a "spam filter" that you could train as you fly, but it turned out to give a lot of both false negatives and false positives. 

So as it currently works, I added controller support that allows you to easily map buttons on your throttle, yoke or joystick to play/pause and to skip. That seems to work quite well until I think of some better solution to automatically skip uninteresting articles.

**Comment by bledo22** [comment]
Beforehand thanks for the response and the add-on. The way I'm using it is in a second PC next to me where I also run LittleNavMap. My main issue at the moment is to not be able to run the moving map on the second PC. I can run vfrmap there but with this version whenever I log to the Map site it obviously says that the sim is not running... I don't really use the speech part, instead read the Wiki article (having a skip button there is useful when you enter a heavily populated area). An in-game pop-up window would be just perfect, but I assume that's a whole different thing.
To sum up... Can you run it on a second PC?, and would also want to know how to map the controls on my joystick...

**Comment by kivle** [comment]
That should be a very simple thing to add as a configuration setting. vfrmap.exe really just spins up a simple webserver that the web app connects to. I can make the url for it configurable in the settings. Should be a very simple change.

**Comment by bledo22** [comment]
Well that would be excellent. I have used vfrmap on my second PC from time to time and it works just fine.

**Comment by kivle** [comment]
I'm having a look at it right now. Like usual with any web app project, if you haven't touched it for a while everything breaks when you try to start it and develop on it again. But it looks like it shouldn't be too much work. A couple of package upgrades and such and I should be able to push out a new version with a setting for the websocket connection. I can't guarantee when I get the time to do it, but I'll ping here when it's ready for testing. :)

**Comment by kivle** [comment]
I have just pushed a new version. It adds a setting where you can override the url to the websocket server. I have only done some minimal testing, but it should work. Tell me how it goes. If there's problems with it, I will have another look.

**Comment by bledo22** [comment]
I've been testing it for two days and I can say that it works wonderfully, no issues whatsoever. Thanks a lot for such a great addon!

**Comment by vector090** [comment]
Hi,
I found your application after using bushtalkradio.com for a while, and think is has advantage in some way, while having some question to ask in the end:

> 1. In a lot of areas on the globe (eg. Africa, etc.) there simply isn't many geotagged articles, so you might be flying for an hour before something shows up.

I myself am mostly interested in visiting hand-crafted landmarks in MSFS (why waste time in AI-faked things), so finding those good POIs won't be a major problem, as long as I figure out what and where they are, by checking out info from offical doc or scenery addon intro.


> 2. In other areas (like the UK for instance), the density of geotagged articles is simply too large. If you fly over London, it will find 3-4 geotagged articles for every few kilometers you fly. So you end up with a huge backlog of articles that the TTS simply does not get the time to read. It will constantly keep reading new articles but the backlog keeps growing and growing.

I use slew mode for sight-seeing, so I find and reach target POI fast, and then study it in paused way. 
Therefor I do have time to read or listen to the wiki.


> 3. In some areas in for instance central Europe there are a lot of "stub" articles for little villages. So you might fly over a small area and get 10 articles that are just "Village X is located in region Y of country Z. It has a population of C". That can get pretty boring, so it would be nice to find a way to filter these out.
> 4. When it searches for articles, the application currently only requests the summary. So it will only retrieve the first 100-200 words of an article. Generally the introduction to the full Wikipedia article. In most cases that's good enough, but in a lot of cases the more interesting stuff might be further down in the text of a Wikipedia article. But if it were to request the full article text, it would need some way of picking out only the most interesting information in an automated way.

As for item 3 and 4, I think that's the phylosophy guiding bushtalkradio.com :
POIs in bushtalkradio totally rely on UGC: all of them are added by human user. 
The contents can be customized, like adding the "interesting" stuff. But it also supports import from wikipedia directly.  So the final quality totally depends on the user himself. 

A quick summary about BushTalkRadio:
= Pros:
Contents are customizable, by human.
Auto display wiki info and play audio when plane arrives at POI.
Local radios, etc.

= Cons:
Content is defined freely and quality could be uncontrolled.
POI trigger radius minimum is 1 KM (!), and is also definable by user.  - in areas where there are lots of POIs, or submitter specified unreasonable large radius, you just can't get a focused info on the POI in front of your eyes, instead you could be hearing introduction about another POI quite far away, which is stupid.
The site seems not maintained.

The cons finally cause me to find another solution. And here I was lead.

After first trial, my opinion about this application is:
= Pros:
Contents all from wiki, more complete than human edition. (can you imagine POIs in Manhattan island only counts a little more ten? That's BushTalkRadio)

= Cons:
- The searching behaviour seems different from the description and the youtube video: I used https://kivle.github.io/msfs-map/, and seems the primary search is not completely based on my current location, but current+radius distance away, which is strange. This phenomenom becomes even more obvious when my plane goes on.
- I only want to search POI under my nose, so I wish to set the search radius to 100 meters. But I don't see the option. Is it removed in latest version?

If the above two issues are clarified, I think I will prefer using msfs-map for the future virtual tours.
Thanks.

**Comment by kivle** [comment]
Well, the application has been designed for flying, and not for use with the drone really. My main use case when I made this was to create flight plans in a normal way and then fly them while seeing what would show up from Wikipedia, and maybe change my course if something interesting showed up. The way you use it might not work great in all cases.

The reason why I have a large search radius is because I try to limit my amount of calls to the Wikipedia APIs to not spam/overload them. Currently it will at most search once every 20 seconds, and it will only search if your airplane has moved at least 1 km from the last position that was searched. You can sort of see the search algorithm here from the code:

https://github.com/kivle/msfs-map/blob/b923d18d585b994157def0c8e3a10700c354f919/src/features/wikipedia/hooks.js#L18

Another thing that might seem confusing at first is that I do not search in a radius around the plane, unless the airplane is going at a speed below 50 knots. If your speed is above 50 knots, it instead searches only in front of the plane in the direction it is flying. I found that it was wasteful to constantly search behind you in places that you have already passed while flying. The calculation for this can be found in this function:

https://github.com/kivle/msfs-map/blob/b923d18d585b994157def0c8e3a10700c354f919/src/features/wikipedia/wikipediaSlice.js#L278

It is currently not possible to change the search radius. It has been hard coded to 10 km. I used to have a dropdown for it, but it had a bug. And since I never changed it myself I just removed it instead. The articles will still be sorted by distance from you, so you will only see things that are in front of you and sorted by closest distance.

If your main use case is to fly with the drone around POIs, maybe simply using something like https://wiki-map.com/ better fits your use case.

**Comment by vector090** [comment]
Thanks for the reply and the open source. I've been trying some tweaks for various situations and so far so good.
